### PR TITLE
NCIP-17

### PR DESCRIPTION
**Pull request by longfin**

### CODE REVIEW

File: NCIP/ncip-17.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  - Also, we must perform a hard-fork for that action changes, per each revision of reward policy, regardless its complexity.
+
+To eliminate fairness and complexity concern, this proposal suggests more simplified reward scheme that, only applied `A` if users saw and staked upon `A`, not `B` (or any other future changes).
+
+# Specification
+
+To simplify, `ClaimStakeReward` action should know reward policy for each users. specifically, the following changes are required:
+
+## `Stake`
+`Stake` action need to record the reward policy at the time of staking.To accomplish this, we need to keep reward policy as immutable and dedicate state address per reward policy.
```
Comment (boscohyun) [inline NCIP/ncip-17.md:35]:
```suggestion
`Stake` action need to record the reward policy at the time of staking. To accomplish this, we need to keep reward policy as immutable and dedicate state address per reward policy.
```

**Review by boscohyun — APPROVED** [review]

File: NCIP/ncip-17.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+- `StakeRegularFixedRewardSheetTableName` (`string`)
+- `StakeRegularRewardSheetTableName` (`string`)
+- `RewardInterval` (`long`)
+- `lockupInterval` (`long`)
+
+## `ClaimStakeReward`
+
+![image](https://github.com/planetarium/NCIPs/assets/128436/7a291498-209e-41dc-a9fa-3efa705916a9)
+
+`ClaimStakeReward` action need to be processed based only recorded policy, without any other aggregation.
```
Comment (moreal) [inline NCIP/ncip-17.md:50]:
```suggestion
`ClaimStakeReward` action needs to be processed based only recorded policy, without any other aggregation.
```

File: NCIP/ncip-17.md (hunk 1 above)
Comment (moreal) [inline NCIP/ncip-17.md:35]:
```suggestion
`Stake` action needs to record the reward policy at the time of staking.To accomplish this, we need to keep reward policy as immutable and dedicate state address per reward policy.
```

File: NCIP/ncip-17.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+To eliminate fairness and complexity concern, this proposal suggests more simplified reward scheme that, only applied `A` if users saw and staked upon `A`, not `B` (or any other future changes).
+
+# Specification
+
+To simplify, `ClaimStakeReward` action should know reward policy for each users. specifically, the following changes are required:
+
+## `Stake`
+`Stake` action need to record the reward policy at the time of staking.To accomplish this, we need to keep reward policy as immutable and dedicate state address per reward policy.
+
```
Comment (longfin) [inline NCIP/ncip-17.md:36]: We should mention about `StakeStateV2`

File: NCIP/ncip-17.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+- `lockupInterval` (`long`)
+
+## `ClaimStakeReward`
+
+![image](https://github.com/planetarium/NCIPs/assets/128436/7a291498-209e-41dc-a9fa-3efa705916a9)
+
+`ClaimStakeReward` action need to be processed based only recorded policy, without any other aggregation.
+
+# Backward Compatibility
+
```
Comment (longfin) [inline NCIP/ncip-17.md:53]: We also need to introduce migration progress for previous staking users here.

**Comment by longfin** [comment]
Replaced by #9 (thanks @moreal )

**Pull request closed without merging**